tokio/src/io/util/mem.rs

```diff
@@ -185,6 +185,7 @@ impl AsyncRead for Pipe {
         cx: &mut task::Context<'_>,
         buf: &mut ReadBuf<'_>,
     ) -> Poll<std::io::Result<()>> {
+        ready!(poll_proceed_and_make_progress(cx));
```

Comment:
We shouldn't make progress if it returns `Poll::Pending`.

Reply:
Yea, looks like instead of calling `make_progress` always, we need to poll proceed at the start, and only call `make_progress` in the two branches below that would return `Poll::Ready`.